distributed/scheduler.py

```diff
@@ -969,40 +1014,56 @@ def __init__(self, name: str):
             self.duration_average = -1
         self.max_exec_time = -1
         self.suspicious = 0
+        self._groups = {}
 
     def add_exec_time(self, duration: float) -> None:
         self.max_exec_time = max(duration, self.max_exec_time)
         if duration > 2 * self.duration_average:
             self.duration_average = -1
 
     def add_duration(self, action: str, start: float, stop: float) -> None:
+        super().add_duration(action, start, stop)
         duration = stop - start
-        self.all_durations[action] += duration
         if action == "compute":
             old = self.duration_average
             if old < 0:
                 self.duration_average = duration
             else:
                 self.duration_average = 0.5 * duration + 0.5 * old
 
+    def transition(self, old: TaskStateState, new: TaskStateState) -> None:
+        super().transition(old, new)
+        self.state_counts[new] += 1
+
+    def add_group(self, tg: TaskGroup) -> None:
+        self._groups[tg] = None
+
+    def remove_group(self, tg: TaskGroup) -> None:
+        # This is important, we need to adjust the stats
+        self._groups.pop(tg)
+        for state, count in tg.states.items():
+            self.states[state] -= count
+        self.duration -= tg.duration
+        self.nbytes_total -= tg.nbytes_total
+        for typename, count in tg._types.items():
+            self._types[typename] -= count
+            if self._types[typename] == 0:
+                del self._types[typename]
+
     @property
-    def states(self) -> dict[TaskStateState, int]:
-        """The number of tasks in each state,
-        like ``{"memory": 10, "processing": 3, "released": 4, ...}``
-        """
-        return merge_with(sum, [tg.states for tg in self.groups])
+    @_deprecated(use_instead="groups")  # type: ignore[misc]
```

Comment:
The fact that we remove groups from prefixes once all their tasks have been forgotten in https://github.com/dask/distributed/blob/5708bdf37fe8daf697115cd6c90c07148b6dffb1/distributed/scheduler.py#L2031-L2035
means that the additional filtering applied to active in https://github.com/dask/distributed/blob/5708bdf37fe8daf697115cd6c90c07148b6dffb1/distributed/scheduler.py#L995-L1001is superfluous.